File: agent-adapter/python/turboflow_adapter/backend.py
"""Abstract backend class — Strategy pattern."""

from __future__ import annotations

from abc import ABC, abstractmethod

from turboflow_adapter.logger import get_logger
from turboflow_adapter.types import (
    BackendCapabilities,
    BackendInfo,
    ExecOptions,
    ExecResult,
    HealthCheckResult,
    McpServer,
)


class AgentBackend(ABC):
    """Base class for all agent backends. Each backend implements this interface."""

    def __init__(self, id: str) -> None:
        self.id = id
        self.log = get_logger(f"tf-adapter.backend.{id}")

    @property
    @abstractmethod
    def name(self) -> str: ...

    @property
    @abstractmethod
    def description(self) -> str: ...

    @property
    @abstractmethod
    def url(self) -> str: ...

    @property
    @abstractmethod
    def license(self) -> str: ...

    @property
    @abstractmethod
    def capabilities(self) -> BackendCapabilities: ...

    # ── Core operations ──────────────────────────────────────────────────

    @abstractmethod
    async def exec(self, prompt: str, options: ExecOptions | None = None) -> ExecResult: ...

    @abstractmethod
    async def version(self) -> str | None: ...

    @abstractmethod
    async def is_installed(self) -> bool: ...

    @abstractmethod
    async def health_check(self) -> HealthCheckResult: ...

    @abstractmethod
    async def install(self) -> None: ...

    # ── MCP operations (optional) ────────────────────────────────────────

    async def mcp_add(self, server: McpServer) -> None:
        self.log.warning("MCP not supported by %s", self.name)

    async def mcp_remove(self, name: str) -> None:
        self.log.warning("MCP not supported by %s", self.name)

    async def mcp_list(self) -> list[McpServer]:
        return []

    # ── Model resolution ─────────────────────────────────────────────────

    @abstractmethod
    def resolve_model(self, model: str) -> str: ...

    # ── Helpers ──────────────────────────────────────────────────────────

    def has_capability(self, cap: str) -> bool:
        return getattr(self.capabilities, cap, False)

    def to_info(self) -> BackendInfo:
        return BackendInfo(
            id=self.id,
            name=self.name,
            description=self.description,
            url=self.url,
            license=self.license,
            capabilities=self.capabilities,
        )

File: agent-adapter/ts/src/backend.ts
// =============================================================================
// TurboFlow Agent Adapter — Abstract Backend (Strategy Pattern)
//
// Each backend implements this interface. The adapter dispatches to the
// active backend without knowing the implementation details.
// =============================================================================

import type {
  BackendCapabilities,
  BackendId,
  ExecOptions,
  ExecResult,
  HealthCheckResult,
  McpServer,
} from './types.js';
import { Logger } from './logger.js';

export abstract class AgentBackend {
  protected readonly log: Logger;

  constructor(protected readonly id: BackendId) {
    this.log = new Logger(`backend:${id}`);
  }

  /** Human-readable name */
  abstract readonly name: string;

  /** Short description */
  abstract readonly description: string;

  /** Project URL */
  abstract readonly url: string;

  /** License identifier */
  abstract readonly license: string;

  /** What this backend supports */
  abstract readonly capabilities: BackendCapabilities;

  // ── Core operations ─────────────────────────────────────────────────────

  /** Execute a prompt through the agent */
  abstract exec(prompt: string, options?: ExecOptions): Promise<ExecResult>;

  /** Get the agent version string */
  abstract version(): Promise<string | null>;

  /** Check if the agent is installed and reachable */
  abstract isInstalled(): Promise<boolean>;

  /** Full health check */
  abstract healthCheck(): Promise<HealthCheckResult>;

  /** Install the agent backend */
  abstract install(): Promise<void>;

  // ── MCP operations (optional) ───────────────────────────────────────────

  /** Register an MCP server. No-op if MCP not supported. */
  async mcpAdd(server: McpServer): Promise<void> {
    this.log.warn(`MCP not supported by ${this.name}`);
  }

  /** Remove an MCP server. No-op if MCP not supported. */
  async mcpRemove(name: string): Promise<void> {
    this.log.warn(`MCP not supported by ${this.name}`);
  }

  /** List registered MCP servers. Empty if not supported. */
  async mcpList(): Promise<McpServer[]> {
    return [];
  }

  // ── Model resolution ────────────────────────────────────────────────────

  /**
   * Resolve a TurboFlow model tier name (opus/sonnet/haiku) to a
   * backend-specific model string. Pass-through for unknown names.
   */
  abstract resolveModel(model: string): string;

  // ── Helpers ─────────────────────────────────────────────────────────────

  /** Check if a specific capability is supported */
  hasCapability(cap: keyof BackendCapabilities): boolean {
    return this.capabilities[cap];
  }

  /** Get backend info as a plain object */
  toInfo() {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      url: this.url,
      license: this.license,
      capabilities: { ...this.capabilities },
    };
  }
}
